src/core/meshcop/tcat_agent.cpp

```diff
@@ -540,6 +547,97 @@ Error TcatAgent::HandlePing(const Message &aIncomingMessage,
+{
+    Error       error  = kErrorNone;
+    const char *name   = otThreadGetNetworkName(&GetInstance());
+    uint16_t    length = static_cast<uint16_t>(strlen(name));
+    ot::Tlv     tlv;
+
+    VerifyOrExit(Get<ActiveDatasetManager>().IsCommissioned(), error = kErrorInvalidState);
+    VerifyOrExit(length > 0, error = kErrorInvalidState);
+
+    tlv.SetType(kTlvResponseWithPayload);
+    tlv.SetLength(static_cast<uint8_t>(length));
+    SuccessOrExit(error = aOutgoingMessage.Append(tlv));
+    SuccessOrExit(error = aOutgoingMessage.AppendBytes(name, length));
+
+exit:
+    return error;
+}
+
+Error TcatAgent::HandleGetDeviceId(Message &aOutgoingMessage)
+{
+    uint8_t  deviceId[OT_EXT_ADDRESS_SIZE];
+    uint16_t length = 0;
+    ot::Tlv  tlv;
+    Error    error = kErrorNone;
+
+    static_assert(OT_EXT_ADDRESS_SIZE > OT_TCAT_MAX_VENDORID_SIZE, "Size of TCAT vendor id is greater than extaddr.");
+
+    if (mVendorInfo->mDeviceIds != nullptr)
+    {
+        // If array is not null return first set device ID.
+        for (uint8_t i = 0; mVendorInfo->mDeviceIds[i].mDeviceIdType != OT_TCAT_DEVICE_ID_EMPTY; i++)
+        {
+            length = mVendorInfo->mDeviceIds->mDeviceIdLen;
+            memcpy(deviceId, mVendorInfo->mDeviceIds[i].mDeviceId, length);
+            break;
+        }
+    }
+
+    if (length == 0)
+    {
+        // If array is null or contains only empty entries return EUI64.
+        otPlatRadioGetIeeeEui64(&GetInstance(), deviceId);
+        length = OT_EXT_ADDRESS_SIZE;
+    }
+
+    tlv.SetType(kTlvResponseWithPayload);
+    tlv.SetLength(static_cast<uint8_t>(length));
+    SuccessOrExit(error = aOutgoingMessage.Append(tlv));
+    SuccessOrExit(error = aOutgoingMessage.AppendBytes(deviceId, length));
+
+exit:
+    return error;
+}
+
+Error TcatAgent::HandleGetExtPanId(Message &aOutgoingMessage)
+{
+    Error                  error = kErrorNone;
+    const otExtendedPanId *name =
+        otThreadGetExtendedPanId(&GetInstance()); // is it possible to check if i't not set ask during review.
+    uint16_t length = OT_EXT_PAN_ID_SIZE;
+    ot::Tlv  tlv;
+
+    VerifyOrExit(Get<ActiveDatasetManager>().IsCommissioned(), error = kErrorInvalidState);
+
+    tlv.SetType(kTlvResponseWithPayload);
+    tlv.SetLength(static_cast<uint8_t>(length));
+    SuccessOrExit(error = aOutgoingMessage.Append(tlv));
+    SuccessOrExit(error = aOutgoingMessage.AppendBytes(name, length));
+
+exit:
+    return error;
+}
+
+Error TcatAgent::HandleGetProvisioningUrl(Message &aOutgoingMessage)
+{
+    Error    error  = kErrorNone;
+    uint16_t length = static_cast<uint16_t>(strlen(mVendorInfo->mProvisioningUrl));
+    ot::Tlv  tlv;
+
+    VerifyOrExit(length != 0, error = kErrorInvalidState);
+
+    tlv.SetType(kTlvResponseWithPayload);
+    tlv.SetLength(static_cast<uint8_t>(length));
+    SuccessOrExit(error = aOutgoingMessage.Append(tlv));
+    SuccessOrExit(error = aOutgoingMessage.AppendBytes(mVendorInfo->mProvisioningUrl, length));
```

Comment:
just a question here: if line 634 TLV-append succeeds, but the AppendBytes here fails, it looks like we get a partially formed response in `aOutgoingMessage` which then gets an additional `ResponseWithStatusTlv` appended in line 470. Wouldn't that be a malformed / out-of-spec response?  Or am I missing something?

Reply:
> Wouldn't that be a malformed 

Yes. But then malformed messages should be generally discarded and not sent.

Reply:
I don't think You are missing anything here. This is exactly what's going to happen. This could happen if we run of space to append the url. I don't see this URL has size limitation in the specification, which means that it could exceed whatever space we have. (I updated this code to handle ext tlv as well.)

Reply:
I think any TCAT TLV needs a max size limit: created JIRA https://threadgroup.atlassian.net/browse/BHC-744 for this. I also agree with Abtin that a malformed / non-specified response should never be sent. If there's not enough memory to assemble the message, for example, then one of  the error codes needs to be returned. This would be e.g. 0x04 General Error if the Commissioner's message is fully correct, or 0x03 Value Error if the Commissioner sent too-long TLVs.